### ISSUE TITLE
Feature request: take/drop n-th argument of list

### ISSUE THREAD

**Issue by t-hahn** [vermaseren/form#86]
Already several times I've come across the situation that I wanted to take (or drop) the n-th argument of a function, where n is known at run-time only.  For an n known at compile-time one could use the ... operator of the preprocessor, of course.
The syntax I have in mind is something like take_(n, args) or take_(args, n), where 'args' would be used as in nargs_(args).  Likewise, drop_(n, args) or drop_(args, n) would drop the n-th argument.  I have no deep feeling for the names, 'take' and 'drop' happen to be the functions that do that in Mathematica.

Best, Thomas

**Comment by vermaseren** [comment]
I have two options here:
1: make a function select_ which does the following:
      select_(f,3,a1,...,n10) -> f(a3,a1,a2,a4,...,a10)
    after which you can do whatever you want with the arguments.
2: give the cycle option in the transform statement the option of having a dollar variable
    for the number by which it has to cycle. The above would become
      Transform,f,cycle(1,last)=-2,cycle(2,last)=+2;
    in which the -2 and the +2 could be like -$amount and +$amount
    This dollar can be determined in a term by term manner if you have it somewhere in a function
    or whatever.
The second option is far more powerful but requires slightly more thinking.
Of course I can do both, but somehow the name select_ seems to be waiting for something not yet
thought of and I cannot see a good other name.

Jos

**Comment by vermaseren** [comment]
I added the function putfirst_
It works as in the example here: 
    S   a,a1,...,a10;
    CF  f,g;
    L   F = g(a,a1,...,a10);
    id  g(?a) = putfirst_(f,4,?a);
    Print;
    .end
   F =
      f(a3,a,a1,a2,a4,a5,a6,a7,a8,a9,a10);
I will put dollars in the transform statement as soon as time permits. And maybe also one or two
new subcommands.

Jos

**Comment by t-hahn** [comment]
Thanks, the functional form is much better.
Still, with putfirst_ I need two id-statements to get the n-th argument.
I had thought that part_(n, args) was very simple, as it only takes the n-th argument of an array.

Best,  Thomas

**Comment by vermaseren** [comment]
Yes, you need two, but otherwise I needed to build two special functions, because you can use
it for both of the functionalies you asked for and I do not like an explosion of special things, because
usually it means that each special thing means an infinitesimal slowdown of the program for people who
do not use those features.

Jos

**Comment by vermaseren** [comment]
Actually I zipped up the tables a bit and now you can do it in one statement, provided you first
define a table. Try this one:
S   a,a1,...,a10;
CF  f,g;
Table h(1:1,f?(a1?,?a));
Fill h(1) = a1;
L   F = g(a,a1,...,a10);
id  g(?b) = h(1,putfirst_(g,4,?b));
Print;
.end
I will leave the rest to your imagination, but it should be clear that now you really have some
extra power under the hood. And if you like to type less, you can always make a macro of it.

It did need some serious recoding of the table declarations. I tested it as well as I could, but if there
is collateral damage, please let me know.

Jos

**Comment by vermaseren** [comment]
And then, you can also do this:
#define MAX "10"
S   a,a1,...,a`MAX',x,n;
CF  f,g;
#do i = 1,`MAX'
Table,sparse,take`i'(0,a1?,...,a`i'?,?a);
Fill take`i' = a`i';
#enddo
Set take:take1,...,take`MAX';
L   F = g(a1,...,a8)*(<x^1*f(1)>+...+<x^8*f(8)>);
Print +f;
.sort
id  g(?b)*f(n?) = take[n](?b);
Print +f;
.end
This uses the just installed zero-dimensional sparse tables. You can also do it with a one-dimensional
table of course, but then you have to add an index.
Maybe this comes closest to what you had in mind.

Jos

**Comment by t-hahn** [comment]
Jos, of course I don't want to impinge on your ways of designing the FORM user interface, and which function names you use.  The part_ (or drop_) thing was just something I had been looking for a couple of times already and thought it must be quite elemental.  And at least the kind workaround you proposed last I have essentially used in the analytical tensor reduction in FormCalc.
In my understanding the part_ or drop_ functions would act as by-number equivalents of the ... preprocessor operator, i.e. they would extract or cancel an argument of a function by number, not through a pattern matching (at least in Mathematica the latter is - slightly - slower).

Thomas

**Comment by tueda** [comment]
Maybe `take_/drop_/part_/...` functions can be unified into one function, if we could implement a function to be converted to an argument field like (internally using a $-variable, possibly?),
```
f(slice_(1,1,x1,...,x5))                        -> f(x1)
f(slice_(2,1,x1,...,x5))                        -> f(x2)
f(slice_(-1,1,x1,...,x5))                       -> f(x5)
f(slice_(1,3,x1,...,x5))                        -> f(x1,x2,x3)
f(slice_(-3,-1,x1,...,x5))                      -> f(x3,x4,x5)
f(slice_(1,2,x1,...,x5),slice_(4,-1,x1,...,x5)) -> f(x1,x2,x3,x4)
```
although the counterpart to `Drop[list,{n}]` (the last one) would not be super fast. Here I assumed the form of `slice_(start,count,args...)` and a negative value for `start` means an index from the last, while `-1` for `count` just means *until the end*.

**Comment by t-hahn** [comment]
Excellent idea!
Perhaps the second argument should be not the count but the final position, i.e. f(slice_(2,1,...)) in your notation would become f(slice_(2,3,...)).  That would make the negative integer (for 'until the end') more intuitive.
Or if that is too slow/cumbersome you could adopt a notation as in PostScript, with a rotation number and count, i.e. arg 1 = number of rotations left (pos)/right (neg) and arg 2 = number of arguments to take.  That would mean e.g. f(slice_(0,1,x1,...,x5)) -> f(x1), f(slice_(2,3,x1,...,x5)) -> f(x3,x4,x5).

**Comment by vermaseren** [comment]
Hi,

It should be easy now for anyone to implement a function along these lines.
Just have a look at the routines/files that involve the putfirst_ and make the necessary
modifications.
At the moment I spent already more time on this than I wanted (still a few papers to write).
It is just that the possibilities of having the patterns in the tables more general than they were
looked very attractive and so did the zero-dimensional tables.
I had also been thinking along the lines of extra options in the transform statement. But that would
involve allowing dollar variables in the arguments and again that requires some thinking not to mess
things up. That should be one of the next things once I have time again. If I have a drop/keep there
it would make more sense. As in
  Transform f,drop(2,$x);
in which $x is provided on a term by term basis. If $x=4 you would obtain
     f(a1,a2,a3,a4,a5,a6) -> f(a1,a5,a6)
etc. But as said, I first went after the tables, because sooner or later people would have run into the 
limitations that the old patterns had. And those limitations did not provide error messages. They just gave
wrong results. Without warnings.
Of course, anybody who manages to insert the code into the transform statement is welcome to do so.
It should be affecting only the transform.c file. This stuff is very localized.

Jos

**Comment by vermaseren** [comment]
For converting a function into an argument field you can look at how the reverse_ function is handled
internally. There has to be a check of course that the guy is sitting inside another function, or there
will be no substitution. The question is inside which routine the substitution should take place in
order to be as fast as possible. It would be either Normalize or TestSub. The last has to scan arguments
already and hence it might not slow things down quite as much.

I presume that your last example should have
     -> f(x1,x2,x4,x5)
?

Jos

**Comment by tueda** [comment]
> I presume that your last example should have -> f(x1,x2,x4,x5) ?

Yes, it should be ` f(x1,x2,x4,x5)`, which is the result with dropping `x3`.

> Perhaps the second argument should be not the count but the final position, i.e. f(slice_(2,1,...)) in your notation would become f(slice_(2,3,...)). That would make the negative integer (for 'until the end') more intuitive.

Maybe Yes. A disadvantage would be that even for the case `count` is a constant (like, taking one argument), `end_index` is not a constant.

So, if I adopt a convention `slice_(start,end,args...)` for a range of `[start,end]` with 1-based indices (I feel this is natural in FORM or Mathematica language),
```
f(slice_(1,1,x1,...,x5))                        -> f(x1)
f(slice_(2,2,x1,...,x5))                        -> f(x2)
f(slice_(-1,-1,x1,...,x5))                      -> f(x5)
f(slice_(1,3,x1,...,x5))                        -> f(x1,x2,x3)
f(slice_(-3,-1,x1,...,x5))                      -> f(x3,x4,x5)
f(slice_(1,2,x1,...,x5),slice_(4,-1,x1,...,x5)) -> f(x1,x2,x4,x5)
```

**Comment by tueda** [comment]
With ignoring the speed performance, the following code using zero-dimensional tables may be interesting:
```
CF f;
S x,n,n1,n2;

* Get [1,1]. nargs >= 1.
Table first(f?(x?,?a));
Fill first() = f(x);

* Get [last,last]. nargs >= 1.
Table last(f?(?a,x?));
Fill last() = f(x);

* Get [2,last]. nargs >= 1.
Table rest(f?(x?,?a));
Fill rest() = f(?a);

* Get [1,last-1]. nargs >= 1.
Table most(f?(?a,x?));
Fill most() = f(?a);

* Join two functions.
Table join(f?(?a),f?(?b));
Fill join() = f(?a,?b);

* Rotate left by n. nargs >= 1.
Table roll(n?int_,f?(?a));
Fill roll() =
  + delta_(n)   * f(?a)
  + thetap_(n)  * roll(n-1,join(rest(f(?a)),first(f(?a))))
  + thetap_(-n) * roll(n+1,join(last(f(?a)),most(f(?a))))
;

* Get [1,n]. 1 <= n <= nargs.
Table firstn(n?pos_,f?(?a));
Table firstnimpl(n?pos0_,f?(?a),f?(x?,?b));
Fill firstn() = firstnimpl(n,f,f(?a,dum_));
Fill firstnimpl() =
  + delta_(n)  * f(?a)
  + thetap_(n) * firstnimpl(n-1,f(?a,x),f(?b))
;

* Get the n-th argument. 1 <= n <= nargs.
Table take(n?pos_,f?(?a));
Fill take() = first(roll(n-1,f(?a)));

* Drop the n-th argument. 1 <= n <= nargs.
Table drop(n?pos_,f?(?a));
Fill drop() = roll(1-n,most(roll(n,f(?a))));

* Get [n1,n2]. Negative indices count from the end. 1 <= n1 <= n2 <= nargs.
Table slice(n1?!{0,},n2?!{0,},f?(?a));
Fill slice() =
  + thetap_(n1) * thetap_(n2) * firstn(n2-n1+1,roll(n1-1,f(?a)))
  + thetap_(n1) * thetap_(-n2) * slice(n1,nargs_(?a)+n2+1,f(?a))
  + thetap_(-n1) * thetap_(n2) * slice(nargs_(?a)+n1+1,n2,f(?a))
  + thetap_(-n1) * thetap_(-n2) * slice(nargs_(?a)+n1+1,nargs_(?a)+n2+1,f(?a))
;

L F0  = f(1,...,9);       * f(1,2,3,4,5,6,7,8,9)
L F1  = first(F0);        * f(1)
L F2  = last(F0);         * f(9)
L F3  = rest(F0);         * f(2,3,4,5,6,7,8,9)
L F4  = most(F0);         * f(1,2,3,4,5,6,7,8)
L F5  = roll(0,F0);       * f(1,2,3,4,5,6,7,8,9)
L F6  = roll(2,F0);       * f(3,4,5,6,7,8,9,1,2)
L F7  = roll(-2,F0);      * f(8,9,1,2,3,4,5,6,7)
L F8  = firstn(3,F0);     * f(1,2,3)
L F9  = take(3,F0);       * f(3)
L F10 = drop(3,F0);       * f(1,2,4,5,6,7,8,9)
L F11 = slice(3,3,F0);    * f(3)
L F12 = slice(3,6,F0);    * f(3,4,5,6)
L F13 = slice(3,-4,F0);   * f(3,4,5,6)
L F14 = slice(-7,6,F0);   * f(3,4,5,6)
L F15 = slice(-7,-4,F0);  * f(3,4,5,6)

P +s;
.end
```

**Comment by vermaseren** [comment]
I knew that these new uses of the tables would inspire creative use.
It looks like a nice little library.
Btw, you can also just use
  Fill first = f(x);
It is trained on either.

I think the speed could be much improved if the theta_ and delta_ functions are processed before
the next table substitution. That would mean that some of that stuff should be moved from Normalize
to TestSub. In the case that they have already a numerical argument that should cause not much
delay and might speed up a number of programs (maybe also summer6 with its complicated theta
functions).
In particular the slice should benefit greatly from it. It would probabaly make things fast enough to
leave them like this.

Jos

**Comment by vermaseren** [comment]
I just checked: the theta_ and delta_ functions are kind of preprocessed in TestSub already.
Hence the speed should be quite OK. Of course built-in functions would be faster, but the question is
whether it is worth it. In that case the transform solution might be quite competitive as well.

Jos

**Comment by crmafra** [comment]
On Mon,  9 May 2016 at 12:34:27 -0700, Jos Vermaseren wrote:

> Of course built-in functions would be faster, but the question is
> whether it is worth it. 

I will definitely use Takahiro's functions from now on!

This kind of funcionality is something that I missed and
I am glad that Thomas' request sparked this discussion. So
thanks to everyone involved!

I would not like to induce a "feature creep" [1] into
FORM's development, but I believe that all the functions
that Takahiro wrote deserve a built-in version: first_(),
last_(), rest_() etc. They would be much easier to use without
setting up tables and they seem to follow the spirit of how
FORM works well with manipulating function arguments.

In any case, thank you again Jos, Takahiro and Thomas.

[1] https://en.wikipedia.org/wiki/Feature_creep

**Comment by vermaseren** [comment]
Independent of the discussion about new functions, I added some facilities to the transform
statement:
   Transform f,dropargs(3,5);     drops the 3-5 arguments
   Transform f,selectargs(3,last-2); keeps only the indicated arguments
   Transform f,addargs(first,$num); adds the indicated arguments in which the $num should expand
      to a legal number during execution. This is more or less the inverse of SplitArg.
   Transform f,mulargs(4,last-$num); multiplies the indicated arguments. Inverse of FactArg.
The transform cycle can now use also a dollar for the amount of cycling, and all ranges in the
transform statement can contain $ variables instead of numbers. No calculations allowed though.
